### PR TITLE
cli options mobile, config and no-tty added, configurable kvdb and beeps, bug fix of otp_pref, immediate refresh

### PR DESCRIPTION
**Pull request by 6mxo89sqh**
1. Ability to pass mobile from cli.
2. Ability to specify config
3. New flag called no-tty to avoid inputs
4. Allow to configure KVDB Bucket Id.
5. Allow to turn off beeps.
6. Bug fix otp_pref is accessed before assignment (comes in case token is passed from cli)
7. Immediate refresh on Ctrl + C

### CODE REVIEW

**Comment by bombardier-gif** [comment]
Thanks for the contribution @6mxo89sqh 
I am going to hold off merging anything new till all the bugs are fixed.
Meanwhile I request you to please test your changes extensively, especially the corner cases

(All the authors claim the PR works but usually it works for just 1 happy case, and causes outages). Please bear with me while we stablize the repo

**Comment by 6mxo89sqh** [comment]
I have tested this. In `main` branch, there is a bug that occurs when you pass token via cli. This PR fixes that.

**Comment by 6mxo89sqh** [comment]
@bombardier-gif Please review and merge this. I just retested it and it is working fine.

**Comment by uvmittal** [comment]
how can i try a pr

**Comment by 6mxo89sqh** [comment]
you can download the zip file from [here](https://github.com/6mxo89sqh/covid-vaccine-booking/archive/refs/heads/enhancements.zip).

alternatively, if use git,
```
git clone https://github.com/6mxo89sqh/covid-vaccine-booking.git
git checkout enhancements
```

**Comment by uvmittal** [comment]
what is cli, please tell

**Comment by uvmittal** [comment]
1. Ability to pass mobile from cli.
2. what doesthis do

**Comment by 6mxo89sqh** [comment]
>     1. Ability to pass mobile from cli.
> 
>     2. what doesthis do

You can pass that as program argument like the following. It makes you faster to resume the script.

`python ./src/covid-vaccine-slot-booking.py --mobile <ur_mobile>`

**Comment by uvmittal** [comment]
> > ```
> > 1. Ability to pass mobile from cli.
> > 
> > 2. what doesthis do
> > ```
> 
> You can pass that as program argument like the following. It makes you faster to resume the script.
> 
> `python ./src/covid-vaccine-slot-booking.py --mobile <ur_mobile>`

really needed this'

**Comment by 6mxo89sqh** [comment]
> > > ```
> > > 1. Ability to pass mobile from cli.
> > > 
> > > 2. what doesthis do
> > > ```
> > 
> > 
> > You can pass that as program argument like the following. It makes you faster to resume the script.
> > `python ./src/covid-vaccine-slot-booking.py --mobile <ur_mobile>`
> 
> really needed this'

in case you have config file already generated, you can pass `--no-tty` also. This avoids inputs and proceeds with smart default. (the target is to reach to polling point asap)

**Comment by uvmittal** [comment]
to use this branch 
use
git fetch origin pull/318/head:enhancement
git checkout enhancement

**Comment by akshaydhoka** [comment]
Even with no-tty option I see, it is asking to confirm the details

      minimum_slots   : 1
        pin_code_location_dtls  : []
        refresh_freq    : 10
        search_option   : 2
        start_date      : 2
        vaccine_type    : SPUTNIK V

Proceed with above info? (y/n Default y):
===================================================================================

**Comment by 6mxo89sqh** [comment]
thanks @akshaydhoka . fixed that.

**Comment by uvmittal** [comment]
bug
when booking for 2nd dose first dose appointment is shown

`
Proceed with above info? (y/n Default y):
The following appointments are active! Please cancel them manually first to continue
+-------+---------------+--------------+--------------+--------+------------+-----------+
|   idx | beneficiary   | name         | state_name   |   dose | date       | slot      |
+=======+===============+==============+==============+========+============+===========+
|     1 | XXXXXXX | XXXXXXXXXX | Delhi        |      1 | 22-04-2021 | Afternoon |
+-------+---------------+--------------+--------------+--------+------------+-----------+
`

**Comment by 6mxo89sqh** [comment]
@uvmittal Thanks for raising it as different issue. That one is not related to this PR.

**Comment by uvmittal** [comment]
> @uvmittal Thanks for raising it as different issue. That one is not related to this PR.

can u fix it

**Comment by 6mxo89sqh** [comment]
@bombardier-gif @BumpyKnuckls Can u plz merge this PR?

**Comment by shailesh** [comment]
is this tested enough to be merged?

**Comment by 6mxo89sqh** [comment]
> is this tested enough to be merged?

Yes. All reported issues are fixed (except #401 which is not related to this)

File: src/utils.py [hunk 1]
```diff
@@ -16,28 +17,34 @@
 
 WARNING_BEEP_DURATION = (1000, 5000)
 
-try:
-    import winsound
+kvdb_bucket = os.getenv('KVDB_BUCKET', 'ASth4wnvVDPkg2bdjsiqMN')
```
Comment (Nakul93) [inline src/utils.py:20]: I think you need to update the KVDB API key here

Comment (6mxo89sqh) [reply src/utils.py:20]: @Nakul93, this is not my key. This is the one that is on master right now. And this is configurable by providing this as env variable. 

Ref https://github.com/bombardier-gif/covid-vaccine-booking/blob/main/src/utils.py#L1042

Comment (Nakul93) [reply src/utils.py:20]: Ok so basically we need to change this in the script, I was rather thinking of taking this as a user input

Comment (6mxo89sqh) [reply src/utils.py:20]: #408 is changing that to user input. #425 is talking about removing kvdb.

Comment (6mxo89sqh) [reply src/utils.py:20]: And you don't need to change in this script. you can do one of following:
```bash
export KVDB_BUCKET=abcd
./src/covid-vaccine-slot-booking.py
```
OR
```bash
KVDB_BUCKET=abcd ./src/covid-vaccine-slot-booking.py
```

Comment (Nakul93) [reply src/utils.py:20]: Ok now it works fine

File: src/utils.py [hunk 2]
```diff
@@ -650,11 +669,14 @@ def check_and_book(
...
-                sys.stdout.flush()
-                time.sleep(1)
+            try:
+                for i in range(refresh_freq, 0, -1):
+                    msg = f"No viable options. Next update in {i} seconds... (Press Ctrl + C to refresh immediately. Press Ctrl + C multiple times to exit.)"
+                    print(msg, end="\r", flush=True)
+                    sys.stdout.flush()
+                    time.sleep(1)
+            except KeyboardInterrupt:
+                print("OK. Refreshing...")
```
Comment (Nakul93) [inline src/utils.py:679]: It would have been more neat if this line didn't print over and over again to show the countdown
image

Comment (6mxo89sqh) [reply src/utils.py:679]: Are you using an IDE? In IDE, it shows like that. In terminal, it doesn't print over and over again. (the `\r` is still there, i haven't removed it :) )

Comment (6mxo89sqh) [reply src/utils.py:679]: You can try with master too, it is the same behavior.

Comment (Nakul93) [reply src/utils.py:679]: no in the main branch it works fine. And yes I'm using terminal

Comment (6mxo89sqh) [reply src/utils.py:679]: Screenshot 2021-06-02 at 1 20 19 AM

Comment (Nakul93) [reply src/utils.py:679]: I tried restarting the terminal to check and still it doesn't work.
Anyways this doesn't bother me that much.

**Review by Nakul93 — APPROVED** [review]
Looks good.

**Comment by 6mxo89sqh** [comment]
Thanks @Nakul93 !

> Looks good.

**Comment by Nakul93** [comment]
> @Nakul93 Can you please help list down the changes that you have tested/reviewed?
Tested: 

1. Configurable kvdb(Although I'd prefer that a common ID being used for this repo)
2. no tty and mobile no options
3. immediate refresh (For me the buggy part was printing the 'No viable options' over and over again rest was fine)
4. Force expiry of token after 10 mins

**Comment by AnjunaBoy** [comment]
> > > ```
> > > 1. Ability to pass mobile from cli.
> > > 
> > > 2. what doesthis do
> > > ```
> > 
> > 
> > You can pass that as program argument like the following. It makes you faster to resume the script.
> > `python ./src/covid-vaccine-slot-booking.py --mobile <ur_mobile>`
> 
> really needed this'



> > @Nakul93 Can you please help list down the changes that you have tested/reviewed?
> > Tested:
> 
> 1. Configurable kvdb(Although I'd prefer that a common ID being used for this repo)
> 2. no tty and mobile no options
> 3. immediate refresh (For me the buggy part was printing the 'No viable options' over and over again rest was fine)
> 4. Force expiry of token after 10 mins

Hi @Nakul93 , could you please advise how to make step 2 and 3 run? I am using python, where do I need to add my number in order to skip entering it every time I run the script?
